### PR TITLE
Add Attachment API

### PR DESCRIPTION
**Pull request by jonasfranz**
Depends on and blocks go-gitea/gitea#3478

Adds `Attachments` field to `Release` and improves the `Attachment` model by making it swagger compatible.

## TODO
- [x] Add client functions for getting attachments

### CODE REVIEW

**Comment by lunny** [comment]
LGTM

**Comment by jonasfranz** [comment]
@lunny Could you please re-review this PR because I've added some other functions which includes a file upload.

File: gitea/attachment.go [hunk 1]
```diff
@@ -3,15 +3,101 @@
...
+func (c *Client) GetReleaseAttachment(user, repo string, release int64, id int64) (*Attachment, error) {
+	a := new(Attachment)
+	err := c.getParsedResponse("GET",
+		fmt.Sprintf("/repos/%s/%s/releases/%d/attachments/%d", user, repo, release, id),
+		nil, nil, &a)
+	return a, err
+}
+
+// CreateReleaseAttachment creates an attachment for the given release
+func (c *Client) CreateReleaseAttachment(user, repo string, release int64, file *os.File) (*Attachment, error) {
```
Comment (lafriks) [inline gitea/attachment.go]: Why not io.Reader for file content?

Comment (jonasfranz) [reply gitea/attachment.go]: Because it is required to get the filename (line 65).

Comment (bkcsoft) [reply gitea/attachment.go]: Pass the filename along? (`fi.Name(), file`)

File: gitea/release.go [hunk 2]
```diff
@@ -26,8 +26,9 @@ type Release struct {
 	// swagger:strfmt date-time
 	CreatedAt time.Time `json:"created_at"`
 	// swagger:strfmt date-time
-	PublishedAt time.Time `json:"published_at"`
-	Publisher   *User     `json:"author"`
+	PublishedAt time.Time     `json:"published_at"`
+	Publisher   *User         `json:"author"`
+	Attachments []*Attachment `json:"attachments"`
```
Comment (jonasfranz) [inline gitea/release.go]: Should I use here attachments or assets (in JSON)?

File: gitea/attachment.go [hunk 3]
```diff
@@ -46,23 +46,17 @@ func (c *Client) GetReleaseAttachment(user, repo string, release int64, id int64
 }
 
 // CreateReleaseAttachment creates an attachment for the given release
-func (c *Client) CreateReleaseAttachment(user, repo string, release int64, file *os.File) (*Attachment, error) {
+func (c *Client) CreateReleaseAttachment(user, repo string, release int64, file *io.Reader, filename string) (*Attachment, error) {
 	// Read file to upload
 	fileContents, err := ioutil.ReadAll(file)
```
Comment (lunny) [inline gitea/attachment.go]: So that, `fileContents` is unnecessary. You can just `_, err = io.Copy(writer, file)`

**Review by lunny — CHANGES_REQUESTED** [review]

File: gitea/attachment.go [hunk 4]
```diff
@@ -47,20 +47,14 @@ func (c *Client) GetReleaseAttachment(user, repo string, release int64, id int64
...
-
 	// Write file to body
 	body := new(bytes.Buffer)
 	writer := multipart.NewWriter(body)
 	part, err := writer.CreateFormFile("attachment", filename)
 	if err != nil {
 		return nil, err
 	}
-	part.Write(fileContents)
+	io.Copy(part, file)
```
Comment (lunny) [inline gitea/attachment.go]: The return error should be handled. Sorry, I haven't make it clear on my previous review.

**Review by lunny — CHANGES_REQUESTED** [review]

File: gitea/release.go (hunk 2 above)
Comment (lunny) [reply gitea/release.go]: ref https://developer.github.com/v3/repos/releases/#list-releases-for-a-repository, it should be assets

**Review by lunny — APPROVED** [review]

File: gitea/release.go (hunk 2 above)
Comment (lunny) [reply gitea/release.go]: @JonasFranzDEV this is not resolved.

**Review by lunny — CHANGES_REQUESTED** [review]

**Review by appleboy — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Pull request merged**